stackdriver-core/lib/stackdriver/core/async_actor.rb

```diff
@@ -0,0 +1,347 @@
+
+        return :timeout unless force
+
+        @thread.kill
+        @thread.join
+        :forced
+      end
+
+      ##
+      # Abstract method that the inheriting classes should implement.
+      #
+      # This should be the main task job that will be run asynchronously and
+      # repeatly.
+      def run_backgrounder
+        fail "#{self.class} class should override #run_backgrounder method"
+      end
+
+      ##
+      # @private Cleanup this async job when process exists
+      #
+      def self.register_for_cleanup actor
+        @exit_lock.synchronize do
+          unless @cleanup_list
+            @cleanup_list = []
+            at_exit { run_cleanup }
+          end
+          @cleanup_list.push actor
+        end
+      end
+
+      ##
+      # @private Take this async job off exit cleanup list
+      #
+      def self.unregister_for_cleanup actor
+        @exit_lock.synchronize do
+          @cleanup_list.delete actor if @cleanup_list
+        end
+      end
+
+      ##
+      # @private Cleanup the async job
+      #
+      def self.run_cleanup
+        @exit_lock.synchronize do
+          if @cleanup_list
+            @cleanup_list.shift.async_stop! until @cleanup_list.empty?
+          end
+        end
+      end
+
+      private_class_method :run_cleanup
+
+      private
+
+      ##
+      # @private Helper method to async_stop! to wait for async job to
+      # terminate.
+      #
+      # @return [Boolean] True if async job terminated. False if timeout.
+      #
+      def wait_until_async_stopped timeout = nil
+        ensure_thread
+        deadline = timeout ? ::Time.new.to_f + timeout : nil
+        synchronize do
+          until async_state == :stopped
+            cur_time = ::Time.new.to_f
+            return false if deadline && cur_time >= deadline
+            max_interval = @cleanup_options[:wait_interval]
+            interval = deadline ? deadline - cur_time : max_interval
+            interval = max_interval if interval > max_interval
+            @lock_cond.wait interval
+          end
+        end
+        true
+      end
+
+      ##
+      # @private Constructor to initialize MonitorMixin
+      #
+      def initialize
+        super()
+        @startup_lock = Mutex.new
+        @cleanup_options = {
+          wait_interval: WAIT_INTERVAL,
+          timeout: CLEANUP_TIMEOUT,
+          force: true
+        }
+      end
+
+      ##
+      # @private Wrapper method for running the async job. It requires classes
+      #   that include AsyncActor module to define a run_backgrounder method.
+      #   Then it runs a loop that checks for the state is workable (:running
+      #   or :suspended), which calls the run_backgrounder method. It ensures
+      #   the state variable gets set to :stopped when this method exists.
+      def async_run_job
+        fail "run_backgrounder method not defined" unless
+          respond_to? :run_backgrounder
```

Comment:
`run_backgrounder` is already defined on this module, so it should always be there.